runtime/drivers/clickhouse/olap.go

```diff
@@ -0,0 +1,540 @@
+	case "UINT16":
+		t.Code = runtimev1.Type_CODE_UINT16
+	case "UINT32":
+		t.Code = runtimev1.Type_CODE_UINT32
+	case "UINT64":
+		t.Code = runtimev1.Type_CODE_UINT64
+	case "UINT128":
+		t.Code = runtimev1.Type_CODE_UINT128
+	case "UINT256":
+		t.Code = runtimev1.Type_CODE_UINT256
+	case "FLOAT32":
+		t.Code = runtimev1.Type_CODE_FLOAT32
+	case "FLOAT64":
+		t.Code = runtimev1.Type_CODE_FLOAT64
+	// can be DECIMAL or DECIMAL(...) which is covered below
+	case "DECIMAL":
+		t.Code = runtimev1.Type_CODE_FLOAT64
+	case "STRING":
+		t.Code = runtimev1.Type_CODE_STRING
+	case "DATE":
+		t.Code = runtimev1.Type_CODE_DATE
+	case "DATE32":
+		t.Code = runtimev1.Type_CODE_DATE
+	case "DATETIME":
+		t.Code = runtimev1.Type_CODE_TIMESTAMP
+	case "DATETIME64":
+		t.Code = runtimev1.Type_CODE_TIMESTAMP
+	case "JSON":
+		t.Code = runtimev1.Type_CODE_JSON
+	case "UUID":
+		t.Code = runtimev1.Type_CODE_UUID
+	case "IPV4":
+		t.Code = runtimev1.Type_CODE_STRING
+	case "IPV6":
+		t.Code = runtimev1.Type_CODE_STRING
+	case "OTHER":
+		t.Code = runtimev1.Type_CODE_JSON
+	default:
+		match = false
+	}
+	if match {
+		return t, nil
+	}
+
+	// All other complex types have details in parentheses after the type name.
+	base, args, ok := splitBaseAndArgs(dbt)
+	if !ok {
+		return nil, fmt.Errorf("encountered unsupported clickhouse type '%s'", dbt)
+	}
+
+	switch base {
+	case "DATETIME":
+		t.Code = runtimev1.Type_CODE_TIMESTAMP
+	case "DATETIME64":
+		t.Code = runtimev1.Type_CODE_TIMESTAMP
+	// Example: "DECIMAL(10,20)", "DECIMAL(10)"
+	case "DECIMAL":
+		t.Code = runtimev1.Type_CODE_DECIMAL
+	case "DECIMAL32":
+		t.Code = runtimev1.Type_CODE_DECIMAL
+	case "DECIMAL64":
+		t.Code = runtimev1.Type_CODE_DECIMAL
+	case "DECIMAL128":
+		t.Code = runtimev1.Type_CODE_DECIMAL
+	case "DECIMAL256":
+		t.Code = runtimev1.Type_CODE_DECIMAL
+	case "FIXEDSTRING":
+		t.Code = runtimev1.Type_CODE_STRING
+	case "ARRAY":
+		t.Code = runtimev1.Type_CODE_ARRAY
+		var err error
+		t.ArrayElementType, err = databaseTypeToPB(dbt[6:len(dbt)-1], true)
+		if err != nil {
+			return nil, err
+		}
+	// Example: "MAP(VARCHAR, INT)"
+	case "MAP":
+		fieldStrs := strings.Split(args, ",")
+		if len(fieldStrs) != 2 {
+			return nil, fmt.Errorf("encountered unsupported clickhouse type '%s'", dbt)
+		}
+
+		keyType, err := databaseTypeToPB(strings.TrimSpace(fieldStrs[0]), true)
+		if err != nil {
+			return nil, err
+		}
+
+		valType, err := databaseTypeToPB(strings.TrimSpace(fieldStrs[1]), true)
+		if err != nil {
+			return nil, err
+		}
+
+		t.Code = runtimev1.Type_CODE_MAP
+		t.MapType = &runtimev1.MapType{
+			KeyType:   keyType,
+			ValueType: valType,
+		}
+	case "ENUM":
+		t.Code = runtimev1.Type_CODE_STRING // representing enums as strings for now
+	default:
+		return nil, fmt.Errorf("encountered unsupported clickhouse type '%s'", dbt)
+	}
+
+	return t, nil
+}
+
+// Splits a type with args in parentheses, for example:
+//
+//	`STRUCT("a" INT, "b" INT)` -> (`STRUCT`, `"a" INT, "b" INT`, true)
```

Comment:
looks like we don't support struct but this example suggests we support so may be change this

Reply:
Yeah it was ported from duckDB driver so ended up copying example as well. Will change it.